packages/framework/src/resources/workflow/workflow.resource.test-d.ts

```diff
@@ -0,0 +1,281 @@
+import { describe, expectTypeOf } from 'vitest';
+import { workflow } from '.';
+import { Subscriber } from '../../types';
+
+describe('workflow function types', () => {
+  describe('event types', () => {
+    it('should have the expected subscriber type', () => {
+      workflow('without-schema', async ({ subscriber }) => {
+        expectTypeOf(subscriber).toEqualTypeOf<Subscriber>();
+      });
+    });
+
+    it('should have the expected step functions', () => {
+      workflow('without-schema', async ({ step }) => {
+        expectTypeOf(step).toMatchTypeOf<{
+          email: unknown;
+          sms: unknown;
+          push: unknown;
+          chat: unknown;
+          inApp: unknown;
+          digest: unknown;
+          delay: unknown;
+          custom: unknown;
+        }>();
+      });
+    });
+  });
+
+  describe('without schema', () => {
+    it('should infer an unknown record type in the step function step controls', async () => {
+      workflow('without-schema', async ({ step }) => {
+        await step.email('without-schema', async (controls) => {
+          expectTypeOf(controls).toEqualTypeOf<Record<string, unknown>>();
+
+          return {
+            subject: 'Test subject',
+            body: 'Test body',
+          };
+        });
+      });
+    });
```

Comment:
New test that asserts a missing control schema should result in an unknown record controls data type